### ISSUE TITLE
Large suites: light weight task proxies.

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc#1689]
On of the major limiting factors for very large suites is memory use by task proxies. 

To reduce this we can:
  * reduce the size of each task proxy object
  * reduce the total number of task proxies

**Comment by hjoliver** [comment]
**Reduce size of task proxies**

Task proxy size is dominated by `[runtime]` settings: a large number of items, and potentially a large amount of user-defined scripting, environment variables etc. replicated through the inheritance hierarchy.  This information is parsed from the suite definition at start-up and is carried throughout the life of a proxy object, BUT *it is only ever needed at job submission time*!

Choices: 
 1. Eliminate all duplication of runtime settings that could be shared by task proxies.

 OR

 2. Only load task runtime at job submit time, then immediately forget it again.

(Note that #1428 was a partial 2. - in effect, all runtime info was dropped from succeeded tasks. This demonstrated a factor of six memory reduction in a large suite with lots of runahead, but the implementation there had other issues that I haven't got back to yet).

**Better sharing of runtime settings?**

 * Default settings are now shared, by means of intervening in the dict look-up mechanism (#1500)
 * Any inherited settings should also be shared by inheriting tasks (like defaults, or is storing references to the same data enough?)

However, it should be noted that the minimum size of task runtimes, with all possible sharing of inherited items and defaults, is still potentially very large in a large suite: it is essentially the information in the suite definition under `[runtime]` *after Jinja2 processing*.

**Comment by hjoliver** [comment]
**Load runtime settings only at job submit time?**

IMO this is clearly the right thing to do, given that (a) the minimum size of task runtime - above - information is potentially very large; and (b) it is not needed for any other purpose.

Choices:
 1. During suite parsing, write each task runtime dict to disk.  At job submit time, load it, write the job file, then forget it again. 
 2. Hold the minimal runtime (above) in memory, and work through inheritance again each time before job submission.
 3. Read and parse the whole suite definition off disk again prior to every job submission.

In reverse order:
 3. is clearly a bad idea - it involves a lot more disk I/O than 1. and a lot more processing (parsing the suite) - and a large suite can take a significant amount of time to parse.
 2. is not a good idea - if all inherited data can be shared by the inheriting tasks (i.e. if inheritance does not duplicate the inherited data) then there is no advantage to delaying this processing until the data is needed. 

 The disk I/O is: write once (for each task proxy class) at suite start-up, read once for each task proxy instance at job submit time.  The I/O does not occur in an inner loop, and the amount of I/O is similar to what is already done a

**Comment by hjoliver** [comment]
Possible caveat: can the task runtime conf files be generated incrementally at start-up without loading the entire runtime configuration - without defaults - into memory first, for inheritance processing?  If not, is brief high memory use at start-up that much better than ongoing high memory use?  The disk-based solution would still be much simpler (no need to bother with the extra complexity of ensuring optimal sharing all settings).

**Comment by hjoliver** [comment]
(this is in fact rather an old idea: https://github.com/cylc/cylc/issues/108#issuecomment-10498633)

**Comment by matthewrmshin** [comment]
With your proposed solution 1, are we going to end up with a new small file per task/job? The only concern is that it increases inode usage on the file system, (and some file systems are very unfriendly to lots of small files), but maybe it does not matter.

(On similar note, but unrelated to this issue, perhaps we should move the `job-activity.log` file one level up, as it is shared between jobs of the same task.)

**Comment by hjoliver** [comment]
No, it'll just be one new small file for each task class, created at suite start-up.  They'll get re-used for each task instance/job.

**Comment by matthewrmshin** [comment]
Sounds like a good compromise.

**Comment by hjoliver** [comment]
To get the benefits of reduced task proxy size, we need to avoid the initial memory high water mark caused by parsing the suite in its entirety (which includes all the task proxy runtime info).  Even when this data is garbage collected after writing the new task proxy runtime config files, the memory may not be returned to the OS by the Python interpreter (although it will be re-used internally) - i.e. the external "resident memory size" of the suite daemon may not go down after suite parsing.

So, we have agreed on the following (via email):

**low-memory suite parsing design**

(Noting that when a *process* finishes, all of its memory is released to the OS).

1. Generate Jinja2-processed file in a subprocess (this is necessarily monolithic, but the whole file is just treated as text by Jinja2).

2. Read and parse the processed file, skipping over (ignoring) all lines in the [runtime] section.

3. Read the [runtime] section line by line just to extract the "inherit" item (i.e. not full parsing) for each namespace, then compute the C3 linearization (inheritance order)

4. Use a process pool to generate the new task config files concurrently, one process for each task: read and parse just the [runtime] namespaces
in the inheritance list for the task (and maybe the inheritance can even be done in-place in a single data structure rather than one for each member of the parent list).

5. At job submit time, have the job-submit command (which is in a sub-process) read the task config file, not the suite daemon.

**Comment by hjoliver** [comment]
[meeting]
  * worth trying and profiling
  * some concern about extra files on disk (one per task, not instance)
   * use one db for all these files?

**Comment by hjoliver** [comment]
#3515 (spawn on demand) does not address this issue but will reduce task pool size so much that it may not be relevant anymore. But before closing this we should consider the ideas above for reducing the memory footprint due to parsing the suite at start-up?
